framework/test/unit/modules/dpos_v2/commands/unlock.spec.ts

```diff
@@ -0,0 +1,450 @@
+					networkIdentifier,
+				})
+				.createCommandExecuteContext();
+			await unlockCommand.execute(context);
+			storedData = codec.decode<VoterData>(
+				voterStoreSchema,
+				await voterSubstore.get(transaction.senderAddress),
+			);
+		});
+
+		it('should remove eligible pending unlock from voter substore', () => {
+			expect(storedData.pendingUnlocks).not.toContainEqual(unlockableObject);
+			expect(storedData.pendingUnlocks).not.toContainEqual(unlockableObject2);
+			expect(storedData.pendingUnlocks).not.toContainEqual(unlockableObject3);
+		});
+
+		it('should not remove ineligible pending unlock from voter substore', () => {
+			expect(storedData.pendingUnlocks).toContainEqual(nonUnlockableObject);
+		});
+	});
+
+	describe(`when self-voted punished account waits ${SELF_VOTE_PUNISH_TIME} blocks and waits unvoteHeight + ${WAIT_TIME_SELF_VOTE} blocks since pomHeight`, () => {
+		beforeEach(async () => {
+			await delegateSubstore.setWithSchema(
+				transaction.senderAddress,
+				{
+					...defaultDelegateInfo,
+					name: 'punishedselfvoter',
+					pomHeights: [lastBlockHeight - SELF_VOTE_PUNISH_TIME],
+				},
+				delegateStoreSchema,
+			);
+			unlockableObject = {
+				delegateAddress: transaction.senderAddress,
+				amount: delegate1.amount,
+				unvoteHeight: lastBlockHeight - WAIT_TIME_SELF_VOTE,
+			};
+			nonUnlockableObject = {
+				delegateAddress: transaction.senderAddress,
+				amount: delegate2.amount,
+				unvoteHeight: lastBlockHeight,
+			};
+			await voterSubstore.setWithSchema(
+				transaction.senderAddress,
+				{
+					sentVotes: [
+						{ delegateAddress: unlockableObject.delegateAddress, amount: unlockableObject.amount },
+						{
+							delegateAddress: nonUnlockableObject.delegateAddress,
+							amount: nonUnlockableObject.amount,
+						},
+					],
+					pendingUnlocks: [unlockableObject, nonUnlockableObject],
+				},
+				voterStoreSchema,
+			);
+			context = testing
+				.createTransactionContext({
+					stateStore,
+					transaction,
+					header,
+					networkIdentifier,
+				})
+				.createCommandExecuteContext();
+			await unlockCommand.execute(context);
+			storedData = codec.decode<VoterData>(
+				voterStoreSchema,
+				await voterSubstore.get(transaction.senderAddress),
+			);
+		});
+
+		it('should remove eligible pending unlock from voter substore', () => {
+			expect(storedData.pendingUnlocks).not.toContainEqual(unlockableObject);
+		});
+
+		it('should not remove ineligible pending unlock from voter substore', () => {
+			expect(storedData.pendingUnlocks).toContainEqual(nonUnlockableObject);
+		});
+	});
+
+	describe(`when self-voted punished account does not wait ${SELF_VOTE_PUNISH_TIME} blocks and waits unvoteHeight + ${WAIT_TIME_SELF_VOTE} blocks since pomHeight`, () => {
+		beforeEach(async () => {
+			await delegateSubstore.setWithSchema(
+				transaction.senderAddress,
+				{
+					...defaultDelegateInfo,
+					name: 'punishedselfvoter',
+					pomHeights: [lastBlockHeight - 1],
```

Comment:
just note here if `lastBlockHeight - 1` will make `unvoteHeight` to be within punishment `lastPomHeight < unlockingObject.unvoteHeight + WAIT_TIME_SELF_VOTE` period.